pkg/pkgrepository/app_deploy.go

```diff
@@ -30,13 +30,111 @@ func (a *App) deploy(tplOutput string) exec.CmdRunResult {
 		}
 
-		return kapp.Deploy(tplOutput, a.startFlushingAllStatusUpdates, func(exec.CmdRunResult) {})
+		// TODO: should we inject this somewhere else? - there's some other rules that might interact with it...
+		fmt.Println("sending to kapp deploy: \n", appendRebaseRule(tplOutput))
+		return kapp.Deploy(appendRebaseRule(tplOutput), a.startFlushingAllStatusUpdates, func(exec.CmdRunResult) {})
 
 	default:
 		return exec.NewCmdRunResultWithErr(fmt.Errorf("Unsupported way to deploy"))
 	}
 }
 
+func appendRebaseRule(tplOutput string) string {
+	return tplOutput + `
+---
+apiVersion: kapp.k14s.io/v1alpha1
+kind: Config
+rebaseRules:
+- ytt:
+    overlayContractV1:
+      overlay.yml: |
+        #@ load("@ytt:data", "data")
+        #@ load("@ytt:yaml", "yaml")
+        #@ load("@ytt:json", "json")
+        #@ load("@ytt:overlay", "overlay")
+        #@ load("@ytt:struct", "struct")
+
+        #@ def get_rev(annotations):
+        #@  if hasattr(annotations, "packaging.carvel.dev/repo-pkg-revision"):
+        #@    return int(annotations["packaging.carvel.dev/repo-pkg-revision"])
+        #@  else:
+        #@    return -1
+        #@  end
+        #@ end
+        #@ if not hasattr(data.values.existing.metadata.annotations, "packaging.carvel.dev/package-repository-ref"):
+        #@   msg = "Error: cannot overwrite package " + data.values.existing.metadata.name + " because it was not created by a package repository."
+        #@   print(msg)
+        #@   fail(msg)
+        #@ end
+
+        #@ def filter(d, s):
+        #@   return {x: v for x, v in d.items() if not x in s}
+        #@ end
+
+        #! TODO: next person who adds an annotation to the set of all kapp-controller annotations is gonna have a bad time.
+        #@ def annotations_are_identical(existing_anns, new_anns):
+        #@   carvel_anns = set(["kapp.k14s.io/disable-original",
+        #@                      "kapp.k14s.io/disable-wait",
+        #@                      "packaging.carvel.dev/package-repository-ref",
+        #@                      "kapp.k14s.io/identity",
+        #@                      "kapp.k14s.io/create-strategy"])
+        #@   return filter(existing_anns, carvel_anns) == filter(new_anns, carvel_anns)
+        #@ end
+
+        #@ def labels_are_identical(existing_labels, new_labels):
+        #@   carvel_labels = set(["kapp.k14s.io/app", "kapp.k14s.io/association"])
+        #@   return filter(existing_labels, carvel_labels) == filter(new_labels, carvel_labels)
+        #@ end
+
+        #@ def is_identical(existing, new):
+        #@   if existing.spec.refName != new.spec.refName:
+        #@     return False, "spec.refName"
+        #@   end
+        #@   if existing.spec.version != new.spec.version:
+        #@     return False, "spec.version"
+        #@   end
+        #@   if struct.decode(existing.spec.template) != struct.decode(new.spec.template):
+        #@    return False, "spec.template"
+        #@   end
```

Comment:
how come we arent checking entire spec section?

Reply:
to be honest I thought it was your idea.